airbyte-integrations/connectors/destination-clickhouse-v2/src/main/kotlin/io/airbyte/integrations/destination/clickhouse_v2/write/load/BinaryRowInsertBuffer.kt

```diff
@@ -37,20 +53,48 @@ private val log = KotlinLogging.logger {}
  */
 @SuppressFBWarnings(
     value = ["NP_NONNULL_PARAM_VIOLATION"],
-    justification = "suspend and fb's non-null analysis don't play well"
+    justification = "suspend and fb's non-null analysis don't play well",
 )
 class BinaryRowInsertBuffer(
     val tableName: TableName,
     private val clickhouseClient: Client,
+    private val airbyteSchema: AirbyteType
 ) {
     // Initialize the inner buffer
-    private val schema = clickhouseClient.getTableSchema(tableName.name, tableName.namespace)
+    private val schema: TableSchema =
+        clickhouseClient.getTableSchema(tableName.name, tableName.namespace)
     @VisibleForTesting internal var inner = InputOutputBuffer()
     @VisibleForTesting
     internal var writer = RowBinaryFormatWriter(inner, schema, ClickHouseFormat.RowBinary)
 
+    internal val formattedSchema =
+        if (airbyteSchema is ObjectType) {
```

Comment:
`asColumns()` already does this for you (and also handles some stupid edge cases, e.g. UnionType)